### PR TITLE
Add velocity/position iteration to expose in engine/editor

### PR DESCRIPTION
**Pull request by thetrung**
@JCash  check this out !

### CODE REVIEW

**Comment by JCash** [comment]
Before moving on:
* What problem does this pull request solve?
* Is it required to change this setting at runtime? (i.e. every frame)
* ..or is it enough to add a setting to game.project to control this behavior?

Remember that these two extra functions add extra size to the engine, for a feature most users don't want/need.
Therefore, I suggest going for a project setting approach.

**Comment by dapetcu21** [comment]
Sorry for intervening, but I was just browsing around and read a bit through your commit. A few notes:
1. Why all the getters? You don't really need any of them, especially not the Lua scripting additions. You can just do `sys.get_config("physics.velocity_iteration")` from Lua to get the value if you really need it. They just increase executable size for no good reason.
2. You left a bunch of random comments around.
3. In `dmGameSystem::GetVelocityIteration()` you have an if that does the same thing in both clauses. You should remove the getters anyway, but just pointing this out.

**Comment by thetrung** [comment]
> Before moving on:
> 
>     * What problem does this pull request solve?
> 
>     * Is it required to change this setting at runtime? (i.e. every frame)
> 
>     * ..or is it enough to add a setting to game.project to control this behavior?
> 
> 
> Remember that these two extra functions add extra size to the engine, for a feature most users don't want/need.
> Therefore, I suggest going for a project setting approach.

@JCash  Please keep in mind that this is just for showing what I'm trying to do as we discussed in #sourcecode channel. Having no intention of actual merging.

**Comment by thetrung** [comment]
> Sorry for intervening, but I was just browsing around and read a bit through your commit. A few notes:
> 
>     1. Why all the getters? You don't really need any of them, especially not the Lua scripting additions. You can just do `sys.get_config("physics.velocity_iteration")` from Lua to get the value if you really need it. They just increase executable size for no good reason.
> 
>     2. You left a bunch of random comments around.
> 
>     3. In `dmGameSystem::GetVelocityIteration()` you have an if that does the same thing in both clauses. You should remove the getters anyway, but just pointing this out.
> 
>     4. You changed the default values from 10 to 16. I'd keep it the way it was so as to not cause unexpected performance regressions in other people's games.

@dapetcu21 This is just to test how I may add new functionality to the engine/editor, please don't take it too serious.
But thanks for the hint. 
The increasing in iteration is just to test how I may change this setting from game.project.

File: engine/physics/src/box2d/Box2D/Dynamics/b2World.h [hunk 1]
```diff
@@ -241,6 +247,8 @@ class b2World
 
 	b2Vec2 m_gravity;
 	bool m_allowSleep;
+	int m_velocityIteration;
+	int m_positionIteration;
```
Comment (JCash) [inline engine/physics/src/box2d/Box2D/Dynamics/b2World.h:251]: These aren't needed, since these are passed in as parameters to the b2World::Step() function.

File: engine/physics/src/physics/physics_2d.cpp [hunk 2]
```diff
@@ -31,6 +31,8 @@ namespace dmPhysics
     : m_Worlds()
     , m_DebugCallbacks()
     , m_Gravity(0.0f, -10.0f)
+    , m_PositionIteration(16)
+    , m_VelocityIteration(16)
```
Comment (JCash) [inline engine/physics/src/physics/physics_2d.cpp:35]: Note that for backwards compatibility, these are not the current values.
In physics_2d.cpp, we pass in values 10,10

File: engine/physics/src/physics/physics_2d.cpp [hunk 3]
```diff
@@ -208,6 +212,10 @@ namespace dmPhysics
         }
         Context2D* context = new Context2D();
         context->m_Worlds.SetCapacity(params.m_WorldCount);
+        // Test Physics 2D Adding new values
```
Comment (JCash) [inline engine/physics/src/physics/physics_2d.cpp:215]: "Test" suggests debug comment?

File: engine/gamesys/src/gamesys/scripts/script_physics.cpp [hunk 4]
```diff
@@ -1073,6 +1119,9 @@ namespace dmGameSystem
         {"set_gravity",     Physics_SetGravity},
         {"get_gravity",     Physics_GetGravity},
 
+        {"get_velocity_iteration",     Physics_GetVelocityIteration},
+        {"get_position_iteration",     Physics_GetPositionIteration},
```
Comment (JCash) [inline engine/gamesys/src/gamesys/scripts/script_physics.cpp:1123]: What is the use case of getting these values in Lua?

File: engine/gamesys/src/gamesys/components/comp_collision_object.cpp [hunk 5]
```diff
@@ -1515,6 +1515,32 @@ namespace dmGameSystem
         }
     }
 
+    int GetVelocityIteration(void* _world)
+    {
+        CollisionWorld* world = (CollisionWorld*)_world;
+        if (world->m_3D)
+        {
+            return dmPhysics::GetVelocityIteration(world->m_World2D);//0;
```
Comment (JCash) [inline engine/gamesys/src/gamesys/components/comp_collision_object.cpp:1523]: You mustn't call a 2D function when only the 3D context is set!

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/meta.properties [hunk 6]
```diff
@@ -83,6 +83,14 @@ gravity_y.type = number
 gravity_y.help = world gravity along y-axis, -10 by default (natural gravity)
 gravity_y.default = -10
 
+velocity_iteration.type = number
+velocity_iteration.help = modify amount of iteration for velocity in Box2D physics
+velocity_iteration.default = 16
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/meta.properties:88]: The default is 10, as seen in the physics_2d.cpp.
Always be mindful of backwards compatibility.

File: engine/physics/src/box2d/Box2D/Dynamics/b2World.h [hunk 7]
```diff
@@ -185,6 +185,12 @@ class b2World
 	/// Get the global gravity vector.
 	b2Vec2 GetGravity() const;
 
+	/// Get physics Box2D velocity iteration.
+	int GetVelocityIteration() const;
+
+	/// Get physics Box2D position iteration.
+	int GetPositionIteration() const;
```
Comment (JCash) [inline engine/physics/src/box2d/Box2D/Dynamics/b2World.h:192]: These functions are redundant, as the values are _passed in_ to the b2World, from physics_2d.cpp, so you don't need to "get them" from b2world.

File: engine/physics/src/physics/physics_2d.cpp [hunk 8]
```diff
@@ -1203,6 +1212,16 @@ namespace dmPhysics
         return gravity;
     }
 
+    int GetVelocityIteration(HWorld2D world)
```
Comment (JCash) [inline engine/physics/src/physics/physics_2d.cpp:1215]: Please adhere to the naming convention, i.e. "GetVelocityIteration2D".

File: engine/physics/src/box2d/Box2D/Dynamics/b2World.cpp [hunk 9]
```diff
@@ -911,7 +911,8 @@ void b2World::Step(float32 dt, int32 velocityIterations, int32 positionIteration
 	b2TimeStep step;
 	step.dt = dt;
 	step.velocityIterations	= velocityIterations;
-	step.positionIterations = positionIterations;
+	step.positionIterations = positionIterations;
```
Comment (JCash) [inline engine/physics/src/box2d/Box2D/Dynamics/b2World.cpp:914]: We want an as clean copy of Box2D as possible. Please revert this change.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/meta.properties [hunk 10]
```diff
@@ -83,6 +83,14 @@ gravity_y.type = number
...
+
+position_iteration.type = number
+position_iteration.help = modify amount of iteration for position in Box2D physics
+position_iteration.default = 16
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/meta.properties:92]: These names can be clearer imho.
They sound like a boolean.
Perhaps "num_position_iterations" is clearer?

**Review by JCash — CHANGES_REQUESTED** [review]

File: engine/physics/src/box2d/Box2D/Dynamics/b2World.h (hunk 1 above)
Comment (thetrung) [reply engine/physics/src/box2d/Box2D/Dynamics/b2World.h:251]: That's right, I will remove them.

File: engine/gamesys/src/gamesys/scripts/script_physics.cpp (hunk 4 above)
Comment (thetrung) [reply engine/gamesys/src/gamesys/scripts/script_physics.cpp:1123]: Just to know that I can get out the value correctly after modified in game.project

File: engine/gamesys/src/gamesys/components/comp_collision_object.cpp (hunk 5 above)
Comment (thetrung) [reply engine/gamesys/src/gamesys/components/comp_collision_object.cpp:1523]: yeah, I just test , will remove or replace with calling to get Bullet's iteration instead.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/meta.properties (hunk 6 above)
Comment (thetrung) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/meta.properties:88]: Will set them to 10 in next PR.

File: engine/physics/src/box2d/Box2D/Dynamics/b2World.cpp (hunk 9 above)
Comment (thetrung) [reply engine/physics/src/box2d/Box2D/Dynamics/b2World.cpp:914]: Ok, so no touching to b2World at all

File: CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,11 @@
+cmake_minimum_required(VERSION 3.0.0)
```
Comment (JCash) [inline CMakeLists.txt:1]: We don't use CMake. Please remove

File: build/.cmake/api/v1/query/client-vscode/query.json [hunk 12]
```diff
@@ -0,0 +1 @@
+{"requests":[{"kind":"cache","version":2},{"kind":"codemodel","version":2}]}
```
Comment (JCash) [inline build/.cmake/api/v1/query/client-vscode/query.json:1]: We don't check in temporary files. Please remove.

**Review by JCash — CHANGES_REQUESTED** [review]
This pull request contains a lot of redundant files and is very difficult to read.
Also, the actual changes aren't commented or motivated.

**Comment by JCash** [comment]
Note that this PR also seems to do a LOT of changes, which aren't related to the title "Add velocity/position iteration to expose in engine/editor". This is also something we don't want. Each PR should solve a well defined problem.

Again, please state the actual existing problem, and why this change should be added.

Also, keep in mind that open pull requests take time for us to review, so we will close pull requests that linger on for too long.

**Comment by thetrung** [comment]
Sorry, I forgot to close this PR @JCash , thought it would only inside our fork, LoL. Many thanks on the comments tho !
